File: laundry/wash_wallet_be/resources/js/types/account.ts
import { BaseFilters, Expense, JournalDetail, Outlet, User } from ".";

export interface Account {
    amount: number;
    id: number;
    ownerId: number;
    parentId: number | null;
    outletId: number | null;
    code: string;
    name: string;
    slug?: string;
    type: AccountType;
    typeLabel: string;
    level: 1 | 2 | 3;
    isSystem: boolean;
    isTransactional: boolean;
    children?: Account[];
    parent?: Account;
    expenses: Expense[];
    expenseSources: Expense[];
    journalDetails: JournalDetail[];
    outlet?: Outlet;
    journalDetailsCount: number;
    owner: User;
    createdAt: string;
    updatedAt: string;
}

export type AccountType =
    | "asset"
    | "liability"
    | "equity"
    | "revenue"
    | "expense";

export interface AccountFormData {
    parentId: number | null | "";
    code: string;
    name: string;
    type?: AccountType | "";
    typeLabel?: string;
}

export interface AccountFilters extends BaseFilters {
    search?: string;
    type?: AccountType | null;
    level?: number | null;
    isTransactional?: boolean | null;
}

File: laundry/wash_wallet/packages/wash_wallet_domain/lib/src/entities/account.dart
import 'package:equatable/equatable.dart';

class Account extends Equatable {
  final int id;
  final int ownerId;
  final int? outletId;
  final int? parentId;
  final String code;
  final String name;
  final String? slug;
  final String type;
  final String? subtype;
  final String? accountRole;
  final int level;
  final bool isSystem;
  final bool isTransactional;
  final bool isActive;
  final double balance;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  const Account({
    required this.id,
    required this.ownerId,
    this.outletId,
    this.parentId,
    required this.code,
    required this.name,
    this.slug,
    required this.type,
    this.subtype,
    this.accountRole,
    required this.level,
    this.isSystem = false,
    this.isTransactional = true,
    this.isActive = true,
    this.balance = 0.0,
    this.createdAt,
    this.updatedAt,
  });

  factory Account.fromModel(dynamic model) {
    return Account(
      id: model.id,
      ownerId: model.ownerId,
      outletId: model.outletId,
      parentId: model.parentId,
      code: model.code,
      name: model.name,
      slug: model.slug,
      type: model.type,
      subtype: model.subtype,
      accountRole: model.accountRole,
      level: model.level,
      isSystem: model.isSystem,
      isTransactional: model.isTransactional,
      isActive: model.isActive,
      balance: model.balance,
      createdAt: model.createdAt,
      updatedAt: model.updatedAt,
    );
  }

  @override
  List<Object?> get props => [
    id,
    ownerId,
    outletId,
    parentId,
    code,
    name,
    slug,
    type,
    subtype,
    accountRole,
    level,
    isSystem,
    isTransactional,
    isActive,
    balance,
    createdAt,
    updatedAt,
  ];
}
